ext/sockets/sockets.c

```diff
@@ -1626,20 +1664,205 @@ PHP_FUNCTION(socket_recvfrom)
 				zend_string_efree(recv_buf);
 				RETURN_FALSE;
 			}
-			ZSTR_LEN(recv_buf) = retval;
-			ZSTR_VAL(recv_buf)[ZSTR_LEN(recv_buf)] = '\0';
 
 			if (UNEXPECTED(!if_indextoname(sll.sll_ifindex, ifrname))) {
 				PHP_SOCKET_ERROR(php_sock, "unable to get the interface name", errno);
 				zend_string_efree(recv_buf);
 				RETURN_FALSE;
 			}
 
-			ZEND_TRY_ASSIGN_REF_NEW_STR(arg2, recv_buf);
+			struct ethhdr *e = (struct ethhdr *)ZSTR_VAL(recv_buf);
+			unsigned short protocol = ntohs(e->h_proto);
+			unsigned char *payload;
+
+			zval obj;
+			object_init_ex(&obj, ethpacket_ce);
+			zend_update_property(Z_OBJCE(obj), Z_OBJ(obj), ZEND_STRL("socket"), arg1);
+			zend_update_property_long(Z_OBJCE(obj), Z_OBJ(obj), ZEND_STRL("headerSize"), ETH_HLEN);
+			zend_update_property_long(Z_OBJCE(obj), Z_OBJ(obj), ZEND_STRL("ethProtocol"), protocol);
+
+			switch (protocol) {
+				case ETH_P_IP: {
+					payload = ((unsigned char *)e + ETH_HLEN);
+					struct iphdr a;
+					memcpy(&a, payload, sizeof(a));
+					struct iphdr *ip = &a;
+					size_t tlayer = ip->ihl * 4;
+					size_t totalip = ntohs(ip->tot_len);
+
+					if (tlayer < sizeof(*ip) || totalip < tlayer || totalip < slen) {
```

Comment:
I think the `totalip < slen` condition is inverted. It should be `totalip > slen`

Reply:
This is the only remark I do not get, why totalip should be <= to slen ?

Reply:
My bad :) I mistaken `slen` with the buffer size. It should be `totalip <= arg3` then.

I think it would help to assign arg1-arg6 to variables with more explicit names, once we know what they stand for.